### PR TITLE
Add --listen-address=127.0.0.1 option

### PR DESCRIPTION
**Pull request by rvl**
Relates to #833 

# Overview

The host option follows the specification here: http://hackage.haskell.org/package/streaming-commons-0.2.1.0/docs/Data-Streaming-Network.html#t:HostPreference

# Comments

CLI client can still only connect to localhost.

### CODE REVIEW

File: lib/cli/src/Cardano/CLI.hs [hunk 1]
```diff
@@ -765,6 +767,16 @@ databaseOption = optionT $ mempty
     <> metavar "DIR"
     <> help "use this directory for storing wallets. Run in-memory otherwise."
 
+-- | [--listen-address=HOSTSPEC], default: 127.0.0.1
+hostPreferenceOption :: Parser HostPreference
+hostPreferenceOption = option str $ mempty
+    <> long "listen-address"
```
Comment (KtorZ) [inline lib/cli/src/Cardano/CLI.hs:774]: what about calling this just `host` ?

File: lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+spec = describe "API Server" $ do
+    it "listens on the local interface" $ do
+        withListeningSocket "127.0.0.1" ListenOnRandomPort $ \(port, socket) -> do
+            getSocketName socket `shouldReturn`
+                SockAddrInet (fromIntegral port) (tupleToHostAddress (0x7f, 0, 0, 1))
+
+    it "can listen on any interface" $ do
+        withListeningSocket "0.0.0.0" ListenOnRandomPort $ \(port, socket) -> do
+            getSocketName socket `shouldReturn`
+                SockAddrInet (fromIntegral port) 0
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs]: If we give `withListeningSocket` a bad host, it currently throws hard with something like:

```
cardano-wallet-jormungandr: Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [AI_PASSIVE,AI_ADDRCONFIG], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = <assumed to be undefined>, addrCanonName = <assumed to be undefined>}, host name: Just "patate", service name: Just "8090"): does not exist (Name or service not known)
```

which is also what the CLI will spit and is definitely not the best user experience. I suggest we:

1/ Catch this error in `withListeningSocket` and make it a value on which we can pattern-match
2/ Have one or more "negative" test case here.
3/ In `Cardano.Wallet.{HttpBridge,Jormungandr}.hs`, log that `withListeningSocket` has exited with an error that is human friendly and explain how the host is completely bonkers

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs (hunk 2 above)
Comment (rvl) [reply lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs]: Yes good idea, thanks.

I have also caught the exception which occurs when the user tries to listen on a privileged port but does not have permission. So we could relax the `>= 1024` restriction on port numbers and still have a reasonable error message.

Also there is a deadlock issue in `failWith`, which I have hacked around

File: lib/cli/src/Cardano/CLI.hs (hunk 1 above)
Comment (rvl) [reply lib/cli/src/Cardano/CLI.hs:774]: You can rename it if you like, but "listen address" is a common and well understood term for network server applications. "host" is too vague in my opinion.

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 3]
```diff
@@ -313,19 +319,41 @@ start settings trace socket rndCtx seqCtx = do
...
+ioToListenError :: HostPreference -> Listen -> IOException -> Maybe ListenError
+ioToListenError hostPreference portOpt e
+    | isAlreadyInUseError e =
+        Just (ListenErrorAddressAlreadyInUse (listenPort portOpt))
+    | isPermissionError e =
+        Just ListenErrorOperationNotPermitted
+    | isDoesNotExistError e =
+        Just (ListenErrorHostDoesNotExist hostPreference)
+    | otherwise =
+        Nothing
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:359]: :+1:

File: lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    it "can listen on any interface" $ do
+        withListeningSocket "0.0.0.0" ListenOnRandomPort $ \case
+            Right (port, socket) -> do
+                getSocketName socket `shouldReturn`
+                    SockAddrInet (fromIntegral port) 0
+            Left e -> fail (show e)
+
+    -- assuming there is no host "patate"
+    it "handles bad host name" $ do
+        withListeningSocket "patate" ListenOnRandomPort $ \res ->
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs:33]: :ok_hand: :potato:

File: lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            res `shouldBe` Left (ListenErrorHostDoesNotExist "patate")
+
+    -- assuming we are not running the tests as root
+    it "handles privileged ports" $ do
+        withListeningSocket "127.0.0.1" (ListenOnPort 23) $ \res ->
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs:43]: :ok_hand:

**Review by KtorZ — APPROVED** [review]
LGTM

**Comment by piotr-iohk** [comment]
Problems I encountered:

```
  --listen-address HOST    Specification of which host to the bind API server
                           to. Can be an IPv[46] address, hostname, or '*'.
```
but:

1. `*`
```

$ cardano-wallet-jormungandr serve --listen-address * --genesis-block-hash 4c05c5bb1dfa2848a2a36249b28d97ef8e6c77c7cb1f9626d1c4024c2470a2c2
Invalid argument `data'
```
2. ipv6 - for any address I wasn't able to make it work.
e.g. loopback = `::1/128 ` (should work for loopback, right?)
```
$ cardano-wallet-jormungandr serve --listen-address ::1/128 --genesis-block-hash 4c05c5bb1dfa2848a2a36249b28d97ef8e6c77c7cb1f9626d1c4024c2470a2c2
....
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 10:33:15.18 UTC] Jörmungandr is ready.
[iohk.cardano-wallet.serve:Alert:ThreadId 4] [2019-10-18 10:33:15.18 UTC] Can't listen on host Host "::1/128". It does not exist.
```

and sometimes it just stops without showing any error:
```
$ cardano-wallet-jormungandr serve --listen-address ::1/128 --genesis-block-hash 4c05c5bb1dfa2848a2a36249b28d97ef8e6c77c7cb1f9626d1c4024c2470a2c2
....
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 10:33:15.18 UTC] Jörmungandr is ready.
$
```
Screenshot from 2019-10-18 12-39-06

**Comment by rvl** [comment]
@piotr-iohk Thanks for trying it.

1. Your shell is interpreting `*` as a glob. Escape it with a backslash or use single quotes `'*'`. This is a standard thing and there is nothing we can do about it.

2. For IPv6, use `::1` or `*6`. Perhaps I can copy the HostPreference specification to the manual.

3. "Sometimes it just stops without showing any error". This is the logging shutdown issue #851.

**Comment by piotr-iohk** [comment]
btw.
```
  --listen-address HOST    Specification of which host to the bind API server
                           to. Can be an IPv[46] address, hostname, or '*'.
```
I reckon that by default it is `localhost` (or `127.0.0.1`) but maybe for consistency with other options it could say:
```
  --listen-address HOST    Specification of which host to the bind API server
                           to. Can be an IPv[46] address, hostname, or '*'. 
                           (default: localhost)
```

**Comment by piotr-iohk** [comment]
So:
```
wlo1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.8.100  netmask 255.255.255.0  broadcast 192.168.8.255
        inet6 fe80::90c2:786f:431:b721  prefixlen 64  scopeid 0x20<link>
```
and:
```
cardano-wallet-jormungandr serve --listen-address fe80::90c2:786f:431:b721 --genesis-block-hash 4c05c5bb1dfa2848a2a36249b28d97ef8e6c77c7cb1f9626d1c4024c2470a2c2
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 11:53:28.32 UTC] Running as v2019.10.16
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 11:53:28.32 UTC] Wallet backend server starting...
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 11:53:28.32 UTC] Node is Jörmungandr on testnet
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 11:53:28.34 UTC] Waiting for Jörmungandr to be ready on tcp/8080
cardano-wallet-jormungandr: Network.Socket.bind: invalid argument (Invalid argument)
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-10-18 11:53:28.35 UTC] Jörmungandr is ready.
```

**Comment by rvl** [comment]
@piotr-iohk That's a link-local address. Use a global address instead (starts with 2000-3FFF).
But you have found an error case that's worth handling: "Network.Socket.bind: invalid argument."

**Comment by piotr-iohk** [comment]
> @piotr-iohk That's a link-local address. Use a global address instead...

@rvl, I was expecting it to work too (cause `192.168.0.100` works fine)...

**Comment by rvl** [comment]
The IPv6 equivalent of 192.168.0.100 are "Unique Local Addresses" (FD00::/8). I also don't often configure IPv6 so I had to do some reading to understand the basics.

You can also test in netcat to see which addresses should work:
```
rodney@blue:~ % nc -6 -l fe80::90c2:786f:431:b721 8090                                                                     
nc: Invalid argument
rodney@blue:~ % nc -6 -l fd7e:acd:c69f:0:d951:dc70:62f0:e54c 8090                                                          
hello netcat
```

Anyway I have added error handling for "invalid argument" and rebased.

I also changed the CLI so that it shows the default. Note that there is a subtle difference between `--listen-address=127.0.0.1` and `--listen-address=localhost`. On many systems the latter results in also listening on `::1`. I would like to keep it as just IPv4 by default.

**Review by KtorZ — APPROVED** [review]

**Pull request merged**